dulwich/tests/compat/server_utils.py

```diff
@@ -237,6 +237,54 @@ def test_fetch_full_depth_into_shallow_clone_from_dulwich(self):
         self.assertEqual([], _get_shallow(clone))
         self.assertReposEqual(clone, self._source_repo)
 
+    def test_fetch_from_dulwich_issue_88_standard(self):
+        # Basically an integration test to see that the ACK/NAK
+        # generation works on repos with common head.
+        self._source_repo = import_repo('issue88_expect_ack_nak_server.export')
+        self.addCleanup(tear_down_repo, self._source_repo)
+        self._client_repo = import_repo('issue88_expect_ack_nak_client.export',
+                                        as_working_copy=True)
+        self.addCleanup(tear_down_repo, self._client_repo)
+        port = self._start_server(self._source_repo)
+
+        run_git_or_fail(['pull', self.url(port), 'master',],
+                        cwd=self._client_repo.path)
+        self.assertReposEqual(self._source_repo, self._client_repo)
+
+    def test_fetch_from_dulwich_issue_88_alternative(self):
+        # likewise, but the case where the two repos have no common parent
+        self._source_repo = import_repo('issue88_expect_ack_nak_other.export')
+        self.addCleanup(tear_down_repo, self._source_repo)
+        self._client_repo = import_repo('issue88_expect_ack_nak_client.export',
+                                        as_working_copy=True)
+        self.addCleanup(tear_down_repo, self._client_repo)
+        port = self._start_server(self._source_repo)
+
+        run_git_or_fail(['config', 'user.email', 'user@example.com'],
+                        cwd=self._client_repo.path)
+        run_git_or_fail(['config', 'user.name', 'User Name'],
+                        cwd=self._client_repo.path)
+        run_git_or_fail(['pull', self.url(port), 'master',],
+                        cwd=self._client_repo.path)
+        # the merge commit will be automatic so hashes are different,
+        # however it being a working directory and done with git it has
+        # a rather predictable result, assuming a default strategy.
+        self.assertEqual(len(os.listdir(self._client_repo.path)), 8)
+
+    def test_push_to_dulwich_issue_88_standard(self):
+        # Same thing, but we reverse the role of the server/client
+        # and do a push instead.
+        self._source_repo = import_repo('issue88_expect_ack_nak_client.export')
```

Comment:
This bug was not easy to debug and it needs the very minimum 16 + 1 commits to generate enough `have` lines to trigger the bug in dulwich which in turn cause standard git client to error out on push.  Perhaps if there is a way to automatically convert that into the utility function calls that can reproduce this bug without the features added, because I've already put far more time into this than I ever wanted to.